Add allgather_base as per our discussion re: ProcessGroup interface.

agolynski opened the pull request:
Summary: Introduce ProcessGroup::allgather_base. No implementation yet: plan to add it one PG backend at a time in a follow up.

Test Plan: No functional changes, no tests yet.

Differential Revision: D19290739

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 9e7d7ac0:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=492ca46e71b9f5071a090bc78601c6621f400c9c&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 492ca46e since Jan 06
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

pietern commented on torch/lib/c10d/ProcessGroup.cpp:
```diff
@@ -62,4 +62,15 @@ ProcessGroup::ProcessGroup(int rank, int size) : rank_(rank), size_(size) {
 
+// Q: It might be better not to implement this at all so end users will break
+// build instead of realizing they forgot to implement at a later time?
+// virtual
```
Sure, that sounds good. You'll need throwing implementations for the existing backends though.

agolynski replied:
done

pietern commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -128,6 +128,15 @@ class ProcessGroup {
+  // confusing to use and is intended for internal use only?
+  virtual std::shared_ptr<ProcessGroup::Work> allgather_base(
+      at::Tensor& outputBuffer,
+      at::Tensor& inputTensor,
```
Nit: mix of buffer and tensor names.

agolynski replied:
fixed

pietern commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -128,6 +128,15 @@ class ProcessGroup {
       const AllgatherOptions& opts = AllgatherOptions()) = 0;
 
+  // Q: Should we move it to protected or annotate with _ since it can be
+  // confusing to use and is intended for internal use only?
```
We could expose it for power users. I think it's fine to leave public. Btw, didn't we call it `single` before?

agolynski replied:
I think we agreed on _base during our meeting with @mrshenli

mrshenli commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -128,6 +128,14 @@ class ProcessGroup {
       std::vector<at::Tensor>& inputTensors,
       const AllgatherOptions& opts = AllgatherOptions()) = 0;
 
+  // For implementers ProcessGroup API and advanced users only.
```
implementers **of**?

agolynski replied:
done

mrshenli reviewed (commented):
@pietern this looks OK to me. Is this sufficient to unblock #28068?

pietern commented:
@mrshenli We need to remove the coalesced variant first as well.

mrshenli commented:
Try to understand the minimum requires to unblock #28068

1. The `allgather_coalesced` API is already in the public API (but not in doc yet), is it OK to remove it? Or do we have to replace it with an implementation in c10d/comm.h/cpp?
2. I guess this also applies to the `allreduce_coalesced` API. Do we need to replace that with `allreduce_base` too?

@pietern @agolynski

agolynski commented:
Discussed with @mrshenli: If we just remove allgather_coalesced it'll break python API right now. How about we leave it and add comments not to touch it: "deprecated. do not use it. do not implement it".

Let's do similar exercise with allreduce_coalesced in a follow up?

agolynski commented:
> This pull request was **exported** from Phabricator. Differential Revision: [[D19290739](https://our.intern.facebook.com/intern/diff/D19290739/)](https://phabricator.intern.facebook.com/D19290739)

Fixed broken build

pietern reviewed (approved):
CI failure is benign. Good to go. Thanks, @agolynski!

agolynski committed:
Add allgather_base as per our discussion re: ProcessGroup interface. (#31892)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31892

Introduce ProcessGroup::allgather_base. No implementation yet: plan to add it one PG backend at a time in a follow up.

Test Plan: No functional changes, no tests yet.

Differential Revision: D19290739

fbshipit-source-id: ec91720ef4f1f6aa445836d8153324ae769cdfd1
